engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java

```diff
@@ -390,6 +394,144 @@ public void shouldResolveIncidentDueToInputCollection() {
+                    .done())
+            .deploy();
+    final long processInstanceKey =
+        ENGINE
+            .processInstance()
+            .ofBpmnProcessId("multi-task")
+            .withVariable(INPUT_COLLECTION, List.of(1, 2, 3))
+            .create();
+    final var serviceTask =
+        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .withElementType(BpmnElementType.SERVICE_TASK)
+            .getFirst();
+    final var job = findNthJob(processInstanceKey, 1);
+    final var nextKey = ENGINE.getZeebeState().getKeyGenerator().nextKey();
+    final int index = Math.toIntExact(ENGINE.getLastWrittenPosition(1));
+    ENGINE.stop();
+    RecordingExporter.reset();
+
+    // when
+    final ProcessInstanceRecord sequenceFlow =
+        Records.processInstance(processInstanceKey, "multi-task")
+            .setBpmnElementType(BpmnElementType.SEQUENCE_FLOW)
+            .setElementId("from-task-to-multi-instance")
+            .setFlowScopeKey(processInstanceKey)
+            .setProcessDefinitionKey(
+                deployment.getValue().getProcessesMetadata().get(0).getProcessDefinitionKey());
+    final ProcessInstanceRecord multiInstanceBody =
+        Records.processInstance(processInstanceKey, "multi-task")
+            .setBpmnElementType(BpmnElementType.MULTI_INSTANCE_BODY)
+            .setElementId("multi-instance")
+            .setFlowScopeKey(processInstanceKey)
+            .setProcessDefinitionKey(
+                deployment.getValue().getProcessesMetadata().get(0).getProcessDefinitionKey());
+
+    ENGINE.writeRecords(
+        RecordToWrite.command().key(job.getKey()).job(JobIntent.COMPLETE, job.getValue()));
+    ENGINE.writeRecords(
+        index + 1,
+        RecordToWrite.event().key(job.getKey()).job(JobIntent.COMPLETED, job.getValue()),
+        RecordToWrite.command()
+            .key(serviceTask.getKey())
+            .processInstance(ProcessInstanceIntent.COMPLETE_ELEMENT, serviceTask.getValue()));
+    ENGINE.writeRecords(
+        index + 3,
+        RecordToWrite.event()
+            .key(serviceTask.getKey())
+            .processInstance(ProcessInstanceIntent.ELEMENT_COMPLETING, serviceTask.getValue()),
+        RecordToWrite.event()
+            .key(serviceTask.getKey())
+            .processInstance(ProcessInstanceIntent.ELEMENT_COMPLETED, serviceTask.getValue()),
+        RecordToWrite.event()
+            .key(nextKey)
+            .processInstance(ProcessInstanceIntent.SEQUENCE_FLOW_TAKEN, sequenceFlow),
+        RecordToWrite.command()
+            .key(nextKey + 1)
+            .processInstance(ProcessInstanceIntent.ACTIVATE_ELEMENT, multiInstanceBody));
+    ENGINE.writeRecords(
+        RecordToWrite.command()
+            .variable(
+                VariableDocumentIntent.UPDATE,
+                Records.variableDocument(processInstanceKey, "{\"items\":0}")));
```

Comment:
:wrench: We could use the existing `causedBy()` method instead of introducing a new method for the source record position. 

It produces the same result but we don't need to calculate the positions.

```suggestion
ENGINE.writeRecords(
        RecordToWrite.command().key(job.getKey()).job(JobIntent.COMPLETE, job.getValue()),
        RecordToWrite.event()
            .key(job.getKey())
            .job(JobIntent.COMPLETED, job.getValue())
            .causedBy(0),
        RecordToWrite.command()
            .key(serviceTask.getKey())
            .processInstance(ProcessInstanceIntent.COMPLETE_ELEMENT, serviceTask.getValue())
            .causedBy(0),
        RecordToWrite.event()
            .key(serviceTask.getKey())
            .processInstance(ProcessInstanceIntent.ELEMENT_COMPLETING, serviceTask.getValue())
            .causedBy(2),
        RecordToWrite.event()
            .key(serviceTask.getKey())
            .processInstance(ProcessInstanceIntent.ELEMENT_COMPLETED, serviceTask.getValue())
            .causedBy(2),
        RecordToWrite.event()
            .key(nextKey)
            .processInstance(ProcessInstanceIntent.SEQUENCE_FLOW_TAKEN, sequenceFlow)
            .causedBy(2),
        RecordToWrite.command()
            .key(nextKey + 1)
            .processInstance(ProcessInstanceIntent.ACTIVATE_ELEMENT, multiInstanceBody)
            .causedBy(2),
        RecordToWrite.command()
            .variable(
                VariableDocumentIntent.UPDATE,
                Records.variableDocument(processInstanceKey, "{\"items\":0}")));
```

Reply:
AFAIK, this doesn't work because it doesn't set the `sourceRecordPosition`, but rather the `sourceIndex`. The source record position is then kept -1L for the entire batch writer.

@saig0 How would you overcome this?